pkg/controller/cloud/pvlcontroller_test.go

```diff
@@ -0,0 +1,186 @@
+	return nil, fmt.Errorf("not implemented")
+}
+
+func mockVolumeFailure(err error) *mockVolumes {
+	return &mockVolumes{volumeLabelsError: err}
+}
+
+func mockVolumeLabels(labels map[string]string) *mockVolumes {
+	return &mockVolumes{volumeLabels: labels}
+}
+
+func TestPVLabels(t *testing.T) {
+	ignoredPV := v1.PersistentVolume{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "noncloud",
+			Initializers: &metav1.Initializers{
+				Pending: []metav1.Initializer{
+					{
+						Name: initializerName,
+					},
+				},
+			},
+		},
+		Spec: v1.PersistentVolumeSpec{
+			PersistentVolumeSource: v1.PersistentVolumeSource{
+				HostPath: &v1.HostPathVolumeSource{
+					Path: "/",
+				},
+			},
+		},
+	}
+	awsPV := v1.PersistentVolume{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "awsPV",
+			Initializers: &metav1.Initializers{
+				Pending: []metav1.Initializer{
+					{
+						Name: initializerName,
+					},
+				},
+			},
+		},
+		Spec: v1.PersistentVolumeSpec{
+			PersistentVolumeSource: v1.PersistentVolumeSource{
+				AWSElasticBlockStore: &v1.AWSElasticBlockStoreVolumeSource{
+					VolumeID: "123",
+				},
+			},
+		},
+	}
+
+	testCases := map[string]struct {
+		vol          v1.PersistentVolume
+		volumeLabels *mockVolumes
+	}{
+		"non-cloud PV": {
+			vol:          ignoredPV,
+			volumeLabels: nil,
+		},
+		"no labels": {
+			vol:          awsPV,
+			volumeLabels: mockVolumeLabels(make(map[string]string)),
+		},
+		"cloudprovider returns nil, nil": {
+			vol:          awsPV,
+			volumeLabels: mockVolumeFailure(nil),
+		},
+		"cloudprovider labels": {
+			vol:          awsPV,
+			volumeLabels: mockVolumeLabels(map[string]string{"a": "1", "b": "2"}),
+		},
+	}
+
+	for d, tc := range testCases {
+		correctPatch := false
+		cloud := &fakecloud.FakeCloud{}
+		client := fake.NewSimpleClientset()
+		fakeWatch := watch.NewFake()
+		client.PrependWatchReactor("persistentvolumes", core.DefaultWatchReactor(fakeWatch, nil))
+		client.PrependReactor("patch", "persistentvolumes", func(a core.Action) (handled bool, ret runtime.Object, err error) {
+			patch := a.(core.PatchActionImpl).GetPatch()
+			obj := &v1.PersistentVolume{}
+			json.Unmarshal(patch, obj)
+			if tc.volumeLabels != nil {
+				for k, v := range tc.volumeLabels.volumeLabels {
+					if obj.ObjectMeta.Labels[k] != v {
+						t.Errorf("%s: label %s expected %s got %s", d, k, v, obj.ObjectMeta.Labels[k])
+						return false, nil, nil
+					}
+				}
+			}
+			if obj.ObjectMeta.Initializers != nil {
+				t.Errorf("%s: initializer wasn't removed: %v", d, obj.ObjectMeta.Initializers)
+				return false, nil, nil
+			}
+			correctPatch = true
+			return true, nil, nil
+		})
+
+		stopCh := make(chan struct{})
+		pvlController := NewPersistentVolumeLabelController(client, cloud)
+		pvlController.ebsVolumes = tc.volumeLabels
+
+		go pvlController.Run(1, stopCh)
+		fakeWatch.Add(&tc.vol)
+		time.Sleep(500 * time.Millisecond)
+		if correctPatch != true {
+			t.Errorf("%s: patch operation failure", d)
+		}
```

Comment:
I think with a simple refactoring you wouldn't need the goroutine and any wait/sleeping logic here.

If `addLabels(key string)` (or some other function before it) looked up the `Volume` from `volumeLister`, the volume could be passed to another function which handles all the labelling logic in a more stateless way which only uses the client and volume plugins. It's the function which receives the Volume that can be table-tested here. Then there's no need for the controller/caches/watch/asynchronicity.

All the controller setup, goroutines, and run loop stuff is ultimately machinery to get a Volume into a labelling function- the unit tests here could exercise just that labelling logic, and integration/e2e tests would exercise the overall controller machinery.

If you do want to unit test the controller bits, that would also be possible (separately) if you want, but I'd suggest considering how to do it without sleeping.

Reply:
I did a refactoring, but not around the volumes.  How the volumes and their labels are retrieved will be modified in future versions of this controller.  Instead, I refactored the patch creation into a separate function and test that.  This has the benefit of the same testing but in a way they should be preserved in future revisions of this controller.